lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package storage
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/ChainSafe/gossamer/pkg/trie"
+)
+
+// storageDiff is a structure that stores the differences between consecutive
+// states of a trie, such as those occurring during the execution of a block.
+// It records updates (upserts), deletions, and changes to child tries.
+// This mechanism facilitates applying state transitions efficiently.
+// Changes accumulated in storageDiff can be applied to a trie using
+// the `applyToTrie` method
+type storageDiff struct {
+	upserts        map[string][]byte
+	deletes        map[string]bool
+	childChangeSet map[string]*storageDiff
+	mtx            sync.RWMutex
+}
+
+// newChangeSet initialises and returns a new storageDiff instance
+func newStorageDiff() *storageDiff {
+	return &storageDiff{
+		upserts:        make(map[string][]byte),
+		deletes:        make(map[string]bool),
```

Comment:
use a `map[string]struct{}` then it will behave as a hashset without allocating any size as value, it works with `map[string]interface{}` as well

Reply:
Good tip! will do

Reply:
I was doing the change but I believe in this case is more intuitive to keep using `map[string]bool`

This will simplify comparisons like:

```go
for _, k := range keys {
   if k > string(key) && !currentTx.deletes[k] {
      return allEntries[k]
   }
}
```

instead of doing:

```go
for _, k := range keys {
   if _, deleted := currentTx.deletes[k]; k > string(key) && !deleted {
      return allEntries[k]
   }
}
```
or 

```go
if cs.deletes[key] {
   delete(cs.deletes, key)
}
```

instead of doing:

```go
if _, deleted := cs.deletes[key]; deleted {
   delete(cs.deletes, key)
}
```

In terms of memory `boolean` takes only 1 byte so I think in this case I prefer to prioritize legibility than memory usage.